libvast/src/transform.cpp

```diff
@@ -65,40 +53,100 @@ const std::vector<std::string>& transform::event_types() const {
-  // TODO: Use the fast-path overload if all steps are `arrow_transform_step`s.
-  for (const auto& step : steps_) {
-    auto transformed = step->apply(std::move(x));
-    if (!transformed)
-      return transformed.error();
-    x = std::move(*transformed);
+caf::error transform::add_slice(table_slice&& x) {
+  VAST_DEBUG("add slice data to transform {}", name_);
+  range_.add(x.offset(), x.rows());
+  auto batch = as_record_batch(x);
+  return add(x.offset(), x.layout(), batch);
+}
+
+caf::expected<std::vector<table_slice>> transform::finish_slice() {
+  VAST_DEBUG("retrieving result from {} steps of transform {}", steps_.size(),
+             name_);
+  auto guard = caf::detail::make_scope_guard([this]() {
+    this->clear_slices();
+  });
+  std::vector<table_slice> result{};
+  auto finished = finish();
+  if (!finished)
+    return std::move(finished.error());
+  for (auto& [offset, layout, batch] : *finished) {
+    auto slice = arrow_table_slice_builder::create(batch, layout);
+    slice.offset(offset);
+    if (!range_.contains(slice.offset(), slice.rows())) {
+      return caf::make_error(ec::invalid_result, "transform returned a slice "
+                                                 "with an offset outside the "
+                                                 "input offsets");
+    }
+    result.push_back(slice);
   }
-  return std::move(x);
+  return result;
+}
+
+void transform::clear_slices() {
+  to_transform_.clear();
+  range_.clear();
 }
 
-std::pair<vast::type, std::shared_ptr<arrow::RecordBatch>>
-transform::apply(vast::type layout,
-                 std::shared_ptr<arrow::RecordBatch> batch) const {
-  VAST_DEBUG("applying {} arrow steps of transform {}", steps_.size(), name_);
+caf::error transform::add(vast::id offset, vast::type layout,
+                          std::shared_ptr<arrow::RecordBatch> batch) {
+  VAST_DEBUG("add arrow data to transform {}", name_);
+  to_transform_.emplace_back(offset, std::move(layout), std::move(batch));
+  return caf::none;
+}
+
+caf::error
+transform::process_queue(const transform_step_ptr& step, batch_queue& queue) {
+  caf::error failed{};
+  const auto size = queue.size();
+  for (size_t i = 0; i < size; ++i) {
+    auto [offset, layout, batch] = std::move(queue.front());
+    queue.pop_front();
+    // TODO: Add a private method `unchecked_apply()` that the transformation
+    // engine can call directly to avoid repeating the check.
+    if (std::find(event_types_.begin(), event_types_.end(),
+                  std::string{layout.name()})
+        == event_types_.end()) {
+      // The transform does not change slices of unconfigured event types.
+      queue.emplace_back(offset, std::move(layout), std::move(batch));
+      continue;
+    }
+    auto add_failed = step->add(offset, std::move(layout), std::move(batch));
+    if (add_failed) {
+      failed = add_failed;
+      queue.clear();
+      break;
+    }
+  }
+  // Finish frees up resource inside the plugin.
```

Comment:
This sounds like something that we should do via RAII. Then you also don't need the `failed` variable and can simply return the error returned from the `step`.